teps/0002-custom-tasks.md

```diff
@@ -0,0 +1,392 @@
+
+In this case, we would expect the `Example` CRD type accepts a parameter named `task-params`, expressed by having a field `.spec.params`, conformant to the `ParamSpec` type.
+
+```
+apiVersion: example.dev/v0
+kind: Example
+metadata:
+  name: my-example
+spec:
+  params:
+  - name: task-param
+    default: goodbye
+  someString: "$(params.task-param), old friend."
+```
+
+When run, the `PipelineRun` controller will create a `Run<Example>`, passing in either the `Pipeline`'s parameter default (`hello`) or the `PipelineRun`-specified override value, if present:
+
+```
+apiVersion: tekton.dev/v1alpha1
+kind: Run
+metadata:
+  name: run-<tekton-generated-suffix>
+spec:
+  ref:
+    apiVersion: example.dev/v0
+    kind: Example
+    name: my-example
+  params:
+  - name: task-param
+    value: hello
+```
+
+The Custom Task author's controller will be watching for creations of `Run`s of this type, and can invoke a Tekton-provided Go package to resolve parameters in string-typed fields, using reflection. This will give them a struct value like:
+
+```
+{
+  "spec": {
+    "params": [ ... ],
+    "someString": "hello, old friend"
+  }
+}
+```
+
+### Result Reporting
+
+Custom Task controllers can report any results via the `Run`'s `.status.results` field:
+
+```
+apiVersion: tekton.dev/v1alpha1
+kind: Run
+metadata:
+  name: run-blah-blah
+spec:
+  ...
+status:
+  results:
+  - name: first-name
+    value: Bob
+```
+
+If their custom defined CRD type has a `.results` field of a compatible type, `Pipeline` validation can take that into account when validating inputs and outputs between Tasks (custom or traditional) in a `Pipeline`.
+
+Controllers can report any results, regardless of whether the underlying CRD object declared them.
+
+### Cancellation
+
+To support cancellation of `Run`s, when a `PipelineRun` is cancelled, the `PipelineRun`  controller will attempt to update any ongoing `Run`s' `.spec.status` to `"Cancelled"`, and update `.status.conditions` to signal unsuccessful execution.
+
+A Custom Task author can watch for this status update and take any corresponding actions (e.g., cancel a cloud build, stop the waiting timer, tear down the approval listener).
+
+Supporting cancellation is optional but recommended.
+
+### Timeout
+
+Today, users can specify a timeout for a component `Task` of a `Pipeline` (see [`PipelineTask.Timeout`](https://godoc.org/github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1#PipelineTask)). The `Run` type will specify a `Timeout` field to hold this value when created as part of a `PipelineRun` (or when `Run`s are created directly). Custom Task authors can read and propagate this field if desired, but a Tekton-owned controller will be responsible for tracking this timeout and updating timed out `Run`s to signal unsuccessful execution.
+
+A Custom Task author can watch for this status update and take any corresponding actions (e.g., cancel a cloud build, stop the waiting timer, tear down the approval listener).
```

Comment:
Similar to the `cancel` case, I think that the custom controller should be the sole responsible to write the "Succeeded" condition of the custom task. The pipeline run controller should have a mechanism to tell the custom task controller to stop the execution of a custom task.

Reply:
That's a reasonable stance. How would you like the PipelineRun controller to signal that to custom task controllers? Just treat timeout as a type of cancellation? Set `.spec.status=TimedOut` and handle that differently?

Is there any situation where a custom task controller would care to know that it was cancelled due to timeout specifically? 🤔